h/services/url_migration.py

```diff
@@ -0,0 +1,117 @@
+                    "target_uri": updates["url"],
+                }
+
+                if "document" in updates:
+                    ann_update_data["document"] = transform_document(
+                        updates["document"], updates["url"]
+                    )
+
+                # Add selectors to annotation if there is no selector of the same
+                # type.
+                #
+                # This change is specifically to aid in the migration of ebook
+                # annotations from a chapter/page URL to the containing book.
+                # The information about which chapter/page the annotation refers
+                # to is then moved into selectors.
+                #
+                # See https://github.com/hypothesis/h/issues/7709
+                if new_selectors := updates.get("selectors"):
+                    selectors = copy.deepcopy(ann.target_selectors)
+                    for new_sel in new_selectors:
+                        if not any(s for s in selectors if s["type"] == new_sel.type):
+                            selectors.append(new_sel)
+                    ann.target_selectors = selectors
+
+                # Update the annotation's `target_uri` and associated document,
+                # and create `Document*` entities for the new URL if they don't
+                # already exist.
+                h.storage.update_annotation(
+                    self.request,
+                    ann.id,
+                    ann_update_data,
+                    # Don't update "edited" timestamp on annotation cards.
+                    update_timestamp=False,
+                )
+
+                # Create an event that is the same as the one used when the
+                # annotation is updated via the API. We send it to consumers
+                # manually rather than using `request.notify_after_commit`
+                # because that mechanism is not available if we're being called
+                # from a Celery task. Also event handlers in `h.subscribers`
+                # try to start a new transaction, which we don't want.
+                event = AnnotationEvent(self.request, ann.id, "update")
+
+                # Re-index annotation in Elasticsearch.
+                search_index = self.request.find_service(name="search_index")
+                search_index.handle_annotation_event(event)
+
+                # Send real-time update notifications for clients watching the
+                # annotation or the new URL.
+                try:
+                    realtime_event_data = {
+                        "action": event.action,
+                        "annotation_id": event.annotation_id,
+                    }
+                    self.request.realtime.publish_annotation(realtime_event_data)
+                except RealtimeMessageQueueError as err:
+                    log.warning(
+                        f"Failed to send realtime update after URL migration: {err}"
+                    )
```

Comment:
Hmm :) I think you can probably get away with not bothering to send realtime events. Other admin pages don't.

Reply:
Indeed this is not essential, it was just convenient when testing to see the changes show up in the client immediately.

Reply:
On second thoughts I'd actually recommend leaving this out of the production version. It's inessential as you say and it might even cause a problem to trigger the system to send a whole batch of potentially tens-of-thousands of real-time notifications at once, which is not something it normally does